doc/manpages/ccnls.1
'\" t
.\"     Title: ccnls
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: 01/10/2014
.\"    Manual: \ \&
.\"    Source: \ \& 0.8.2
.\"  Language: English
.\"
.TH "CCNLS" "1" "01/10/2014" "\ \& 0\&.8\&.2" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ccnls \- Attempts to list name components available at the next level of the hierarchy for a given CCNx name prefix\&.
.SH "SYNOPSIS"
.sp
\fBccnls\fR \fIccnxname\fR
.SH "DESCRIPTION"
.sp
The \fBccnls\fR utility lists name components available after the supplied CCNx name prefix, \fIccnxname\fR, for content objects that are cached in the local \fBccnd\fR\&.
.sp
This utility does not have explicit flags or options, but can be modified via environment variables\&. These include the scope of the generated interest (\fICCN_SCOPE\fR), the amount of time to wait between responses from \fBccnd\fR (\fICCN_LINGER\fR) and whether signature verification is desired (\fICCN_VERIFY\fR)\&.
.SH "OPTIONS"
.sp
The \fBccnls\fR utility does not have additional flags or options\&. Behavior can be changed by using the following two environment variables\&.
.sp
.if n \{\
.RS 4
.\}
.nf
CCN_SCOPE=
  Scope in generated interests (0, 1, 2)\&. The default is none specified\&.
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
CCN_LINGER=
  Timeout for not receiving data\&.  The default is 0\&.5 seconds\&.
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
CCN_VERIFY=
  If set and non\-empty, indicates that signature verification is required\&.
.fi
.if n \{\
.RE
.\}
.SH "EXIT STATUS"
.PP
\fB0\fR
.RS 4
Success
.RE
.PP
\fB1\fR
.RS 4
Failure (syntax or usage error; retrieval error)
.RE
.SH "AUTHOR"
.sp
Michael Plass


NEWS
Changes between 0.8.1 and 0.8.2:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/21 for more details on CCNx 0.8.2.

- This release features a refactored strategy layer, making it straightforward to add new
  strategies in a modular way.  Alongside the traditional "default" strategy there are
  two new examples, a "parallel" strategy that sends interests on all available faces
  and a "loadsharing" strategy that sends each arriving interest to just one of the
  available faces.  Strategies may be selected by prefix.  This selection may be
  done using new subcommands to the ccndc utility.

	* Feature #100929: Implement Junit test suites to facilitate bulk testing.
	* Feature #100983: strategy layer refactoring
	* Feature #100984: strategy labels on prefixes
	* Feature #100990: Restore SYNC_NOISY to INFO
	* Feature #100992: Add test-security to standard java tests
	* Feature #100993: Add test-versioning back into standard java tests
	* Feature #100994: Remove unused cruft from build.xml
	* Feature #101036: java sync should use Deque operations instead of Stack as recommended in the javadocs
	* Feature #101039: UDP mode for java netmanagers is not tested in standard testing
	* Feature #101053: ccncatchunks2 could use improved hole-filling and interest lifetime management
	* Bug #101030: build.xml setup of ccnr environment is incorrect
	* Bug #101033: Handling of extremely large values of MaxSuffixComponents is incorrect
	* Bug #101037: ccn_seqwriter library code should not limit the block size to 4096 bytes
	* Bug #101042: CcnFileProxy uses deprecated methods
	* Bug #101046: VLC 2.1.0 release necessitates change in plugin Makefile
	* Bug #101048: Do not rely on accept(2) returned addr for recognizing a unix-domain socket
	* Bug #101052: Unbreak objtree build target
	* Bug #101057: ccn_sign_content / ccn_chk_signing_params do not properly incorporate a timestamp in a template via ccn_signing_params
	* Bug #101058: ccndc in daemon mode does not exit cleanly when ccnd goes away
	* Bug #101059: testTimeouts a little too strict
	* Bug #101060: ccnd cache lookups should take a controllable amount of work
	* Bug #101061: ccnd minimum interest lifetime problem
	* Bug #101062: some ccnd logging missing check for logging enabled before calling expensive log procedure
	* Support #101047: Prepare for ccnx-0.8.2 release

Changes between 0.8.0 and 0.8.1:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/20 for more details on CCNx 0.8.1.

- This release fixes a number of bugs, including a serious memory leak in ccnd.

	* Bug #101029: unbounded test log file growth
	* Bug #101032: ccnr should start at specified segment on checked start write
	* Bug #101034: No obvious documentation for how to start ccnr exists on website
	* Bug #101035: small leak in ccn_fetch should a realloc() fail
	* Bug #101038: Inconsistent encoding of zero-length name components
	* Bug #101043: ccnr r_store_lookup_backwards can fail
	* Bug #101044: valgrind reports memory leak in 0.8.0 ccnd
	* Bug #101049: ccnd can remove newly arrived content too quickly
	* Support #101045 Prepare for ccnx-0.8.1 release

Changes between 0.7.2 and 0.8.0:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/17 for more details on CCNx 0.8.0.

- The use of symmetric key HMACs in place of public/private key signatures
  is now an option in both libraries.  In situations where it can be used,
  it provides a much higher performance alternative.
- The content store in ccnd has been improved.  It now uses a flatname
  representation for indexing, in the same way that ccnr does.  Also, the
  underlying skiplist implementation has moved into the library, so that
  applications may take advantage of it in the future.  In the process,
  some old ccnd bugs relating to the handling of extreme values of
  FreshnessSeconds were fixed.
- The more readable escaping convention in URIs is now the default.
  If for whatever reason you prefer to use only percent-escaping in URIs, use

	CFLAGS='-g -DCCN_URI_DEFAULT_ESCAPE=4' ./configure

  before typing make.  Note that both forms of escaping are accepted as input.
- An example rc startup script is now included, but needs to be installed by
  hand.  It is coded for Ubuntu Linux, but should be easily adaptable to other
  systems.
- This release drops support for Java JDK 1.5.

  * Issues Resolved:
	* Bug #100118: ccnd does not handle FreshnessSeconds=0 correctly.
	* Bug #100119: ccnd should handle large values of FreshnessSeconds
	* Bug #100558: test-security does not pass
	* Bug #100897: timeout values  should be consistently long throughout java code 
	* Bug #100966: Incorrect processing of CCNS_SYNC_SCOPE=3
	* Bug #100977: Problem with verify in CCNInputStream when trying to retrieve key
	* Bug #100987: Non-portable use of local keyword in Bourne shell scripts
	* Bug #100998: VLC 2.1.x block_New definition removed
	* Bug #100999: java BLOCK_AFTER_FIRST_SEGMENT doesn't work in all cases
	* Bug #101000: sync_diff code should reset differencing data when aborting or completing a comparison
	* Bug #101004: QueuedContentHandler has a sync error
	* Bug #101005: CCNx Android Services build out of sync with API changes to CCNx C-lib
	* Bug #101007: ccn_resolve_version() should distinguish between no response and finding an unversioned object.
	* Bug #101010: ccnseqwriter command busy-loops when its ccnd goes away
	* Bug #101011: Compile warning when building new symmetric key code with openssl 0.9.8
	* Bug #101012: Improve recovery from node fetch timeout in java sync
	* Bug #101013: node fetch timeout in java sync implementation should be longer and settable

csrc/cmd/ccnls.c
/**
 * @file ccnls.c
 * Attempts to list name components available at the next level of the hierarchy.
 *
 * A CCNx command-line utility.
 *
 * Copyright (C) 2008, 2009, 2011, 2013 Palo Alto Research Center, Inc.
 *
 * This work is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License version 2 as published by the
 * Free Software Foundation.
 * This work is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
 * for more details. You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 */
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ccn/ccn.h>
#include <ccn/charbuf.h>
#include <ccn/uri.h>

struct upcalldata {
    int magic; /* 856372 */
    long *counter;
    unsigned warn;
    unsigned option;
    int n_excl;
    int scope;
    struct ccn_charbuf **excl; /* Array of n_excl items */
};

#define MUST_VERIFY 0x01

static int /* for qsort */
namecompare(const void *a, const void *b)
{
    const struct ccn_charbuf *aa = *(const struct ccn_charbuf **)a;
    const struct ccn_charbuf *bb = *(const struct ccn_charbuf **)b;
    int ans = ccn_compare_names(aa->buf, aa->length, bb->buf, bb->length);
    if (ans == 0)
        fprintf(stderr, "wassat? %d\n", __LINE__);
    return (ans);
}

enum ccn_upcall_res
incoming_content(
    struct ccn_closure *selfp,
    enum ccn_upcall_kind kind,
    struct ccn_upcall_info *info)
{
    struct ccn_charbuf *c = NULL;
    struct ccn_charbuf *comp = NULL;
    struct ccn_charbuf *uri = NULL;
    struct ccn_charbuf *templ = NULL;
    const unsigned char *ccnb = NULL;
    size_t ccnb_size = 0;
    struct ccn_indexbuf *comps = NULL;
    int matched_comps = 0;
    int res;
    int i;
    struct upcalldata *data = selfp->data;
    
    if (data->magic != 856372) abort();
    if (kind == CCN_UPCALL_FINAL)
        return(CCN_UPCALL_RESULT_OK);
    if (kind == CCN_UPCALL_INTEREST_TIMED_OUT)
        return(CCN_UPCALL_RESULT_REEXPRESS);
    if (kind == CCN_UPCALL_CONTENT_UNVERIFIED) {
        if ((data->option & MUST_VERIFY) != 0)
        return(CCN_UPCALL_RESULT_VERIFY);
        }
    else if (kind != CCN_UPCALL_CONTENT) abort();
    
    ccnb = info->content_ccnb;
    ccnb_size = info->pco->offset[CCN_PCO_E];
    comps = info->content_comps;
    matched_comps = info->pi->prefix_comps;
    c = ccn_charbuf_create();
    uri = ccn_charbuf_create();
    templ = ccn_charbuf_create();
    /* note that comps->n is 1 greater than the number of explicit components */
    if (matched_comps > comps->n) {
        ccn_uri_append(c, ccnb, ccnb_size, 1);
        fprintf(stderr, "How did this happen?  %s\n", ccn_charbuf_as_string(uri));
        exit(1);
    }
    data->counter[0]++;
    /* Recover the same prefix as before */
    ccn_name_init(c);
    res = ccn_name_append_components(c, info->interest_ccnb,
                                     info->interest_comps->buf[0],
                                     info->interest_comps->buf[matched_comps]);
    if (res < 0) abort();
    
    comp = ccn_charbuf_create();
    ccn_name_init(comp);
    if (matched_comps + 1 == comps->n) {
        /* Reconstruct the implicit ContentObject digest component */
        ccn_digest_ContentObject(ccnb, info->pco);
        ccn_name_append(comp, info->pco->digest, info->pco->digest_bytes);
    }
    else if (matched_comps < comps->n) {
        ccn_name_append_components(comp, ccnb,
                                   comps->buf[matched_comps],
                                   comps->buf[matched_comps + 1]);
    }
    res = ccn_uri_append(uri, comp->buf, comp->length, 0);
    if (res < 0 || uri->length < 1)
        fprintf(stderr, "*** Error: ccnls line %d res=%d\n", __LINE__, res);
    else {
        if (uri->length == 1)
            ccn_charbuf_append(uri, ".", 1);
        printf("%s%s\n", ccn_charbuf_as_string(uri) + 1,
               kind == CCN_UPCALL_CONTENT ? " [verified]" : " [unverified]");
    }
    ccnb_element_begin(templ, CCN_DTAG_Interest);
    ccn_charbuf_append(templ, c->buf, c->length); /* Name */
    if (matched_comps == comps->n) {
        /* The interest supplied the digest component */
        ccn_charbuf_destroy(&comp);
        /*
         * We can't rely on the Exclude filter to keep from seeing this, so 
         * say that we need at least one more name component.
         */
        ccnb_append_tagged_udata(templ, CCN_DTAG_MinSuffixComponents, "1", 1);
    }
    else {
        data->excl = realloc(data->excl, (data->n_excl + 1) * sizeof(data->excl[0]));
        data->excl[data->n_excl++] = comp;
        comp = NULL;
    }
    qsort(data->excl, data->n_excl, sizeof(data->excl[0]), &namecompare);
    ccnb_element_begin(templ, CCN_DTAG_Exclude);
    for (i = 0; i < data->n_excl; i++) {
        comp = data->excl[i];
        if (comp->length < 4) abort();
        ccn_charbuf_append(templ, comp->buf + 1, comp->length - 2);
    }
    comp = NULL;
    ccnb_element_end(templ); /* </Exclude> */
    ccnb_tagged_putf(templ, CCN_DTAG_AnswerOriginKind, "%d", CCN_AOK_CS);
    if (data->scope > -1)
       ccnb_tagged_putf(templ, CCN_DTAG_Scope, "%d", data->scope);
    ccnb_element_end(templ); /* </Interest> */
    if (templ->length > data->warn) {
        fprintf(stderr, "*** Interest packet is %d bytes\n", (int)templ->length);
        data->warn = data->warn * 8 / 5;
    }
    ccn_express_interest(info->h, c, selfp, templ);
    ccn_charbuf_destroy(&templ);
    ccn_charbuf_destroy(&c);
    ccn_charbuf_destroy(&uri);
    return(CCN_UPCALL_RESULT_OK);
}

void
usage(const char *prog)
{
    fprintf(stderr, "Usage: %s uri\n"
            "   Prints names with uri as prefix\n"
            "     environment var CCN_SCOPE is scope for interests (0, 1 or 2, no default)\n"
            "     environment var CCN_LINGER is no-data timeout (seconds) default 0.5s\n"
            "     environment var CCN_VERIFY indicates signature verification is required (non-zero)\n", prog);
    exit(1);
}

int
main(int argc, char **argv)
{
    struct ccn *ccn = NULL;
    struct ccn_charbuf *c = NULL;
    struct ccn_charbuf *templ = NULL;
    struct upcalldata *data = NULL;
    int i;
    int n;
    int res;
    long counter = 0;
    struct ccn_closure *cl = NULL;
    int timeout_ms = 500;
    const char *env_timeout = getenv("CCN_LINGER");
    const char *env_verify = getenv("CCN_VERIFY");
    const char *env_scope = getenv("CCN_SCOPE");

    if (argv[1] == NULL || argv[2] != NULL)
        usage(argv[0]);

    if (env_timeout != NULL && (i = atoi(env_timeout)) > 0)
        timeout_ms = i * 1000;

    c = ccn_charbuf_create();
    res = ccn_name_from_uri(c, argv[1]);
    if (res < 0)
        usage(argv[0]);
        
    ccn = ccn_create();
    if (ccn_connect(ccn, NULL) == -1) {
        perror("Could not connect to ccnd");
        exit(1);
    }
    
    data = calloc(1, sizeof(*data));
    data->magic = 856372;
    data->warn = 1492;
    data->counter = &counter;
    data->option = 0;
    if (env_verify && *env_verify)
        data->option |= MUST_VERIFY;
    data->scope = -1;
    if (env_scope != NULL && (i = atoi(env_scope)) >= 0)
      data->scope = i;
    cl = calloc(1, sizeof(*cl));
    cl->p = &incoming_content;
    cl->data = data;
    if (data->scope > -1) {
        templ = ccn_charbuf_create();
        ccnb_element_begin(templ, CCN_DTAG_Interest);
        ccnb_element_begin(templ, CCN_DTAG_Name);
        ccnb_element_end(templ); /* </Name> */
        ccnb_tagged_putf(templ, CCN_DTAG_Scope, "%d", data->scope);
        ccnb_element_end(templ); /* </Interest> */
    }
    ccn_express_interest(ccn, c, cl, templ);
    ccn_charbuf_destroy(&templ);
    cl = NULL;
    data = NULL;
    for (i = 0;; i++) {
        n = counter;
        ccn_run(ccn, timeout_ms); /* stop if we run dry for 1/2 sec */
        fflush(stdout);
        if (counter == n)
            break;
    }
    ccn_destroy(&ccn);
    ccn_charbuf_destroy(&c);
    exit(0);
}
